### ISSUE TITLE
Windows: Files restricted to 64-bit processes are invisible to Chef and recipes

### ISSUE THREAD

**Issue by adamedx** [chef/chef#3409]
Workaround for this issue is below.

On Windows, Chef is executed by the Ruby runtime a 32-bit process. Because of choices made in Windows around backward compatibility with 32-bit applications on 64-bit Windows, certain parts of the file system are "hidden" from 32-bit processes (e.g. places where the path might try to load a 64-bit dll into a 32-bit process which won't work, or other places where configuration must be kept separate to avoid 32-bit code and 64-bit code interacting). What this means is that the following Chef recipe will not function as expected:

```ruby
file 'c:/hello..txt' do
  content 'world'
  only_if { ::File.exist? 'c:/windows/system32/tasks/somefile' }
end
```
This will fail because for the `only_if` guard, Windows will use a different path than `c:/windows/system32/tasks`, it will use `c:/windows/syswow64/tasks` instead.

This redirection can be overridden, and in fact Chef does that when running script resources like `batch` and `powershell_script`. The above limitation then can be worked around using the following guard expression:

```ruby
guard_interpreter :powershell_script
only_if 'test-path "c:/windows/system32/tasks/somefil"'
```
The real fix for this is for Chef to use a 64-bit process :smile:. Other workarounds include adding a primitive to the DSL to make a resource execute with 32-bit redirection disabled, but that could have unpredictable side effects. We could also make the entire chef run work with 32-bit redirection disabled, or at least make it an option, but again this is very risky.

So really we need to get 64-bit Chef -- good thing we've demo'd a preview...

**Comment by adamedx** [comment]
Here is a recipe that demonstrates that 32-bit redirection is the problem -- it uses some internal helpers to disable redirection. Run it with `chef-apply` and the output should be self-explanatory. If you get an exception, that means the recipe is "wrong" or we don't really understand the bug. 

```ruby
file_only_visible_on_64_bit = "#{ENV['SystemRoot']}/system32/Tasks/64OnlyFile".gsub(/\\/, '/')

powershell_script 'make64onlyfile' do
  code "echo hi > #{file_only_visible_on_64_bit}"
end

arch_mode_32_bit_failed = true

ruby_block 'this should not run but it does' do
  block do
    arch_mode_32_bit_failed = true
    puts "\n\n****This ran because Ruby is 32-bit and can't see the 64-bit only file #{file_only_visible_on_64_bit}.\n"
  end

  # We're 32-bit here, so the 64-bit only file can't be seen
  not_if { ::File.exist? file_only_visible_on_64_bit }
end

arch_mode_64_bit_failed = true

# WORKAROUND -- use with_os_architecture for expressions that need 64-bit
require 'chef/mixin/windows_architecture_helper'

ruby_block 'this should run and does because we use a workaround to see the file' do
  block do
    arch_mode_64_bit_failed = false
    #    arch_32_failed(false)
    puts "\n\n****This will run because of the workaround -- the workaround works!\n"
  end

  # We're 64-bit thanks to with_os_architecture, so now
  # the 64-bit file can be seen
  only_if do
    self.class.class_eval('include Chef::Mixin::WindowsArchitectureHelper')
    # Any code that needs to see the 64-bit view of the world should
    # be inside this with_os_architecture block
    with_os_architecture(node) do
      ::File.exist? file_only_visible_on_64_bit # must be executed in 64-bit mode
    end
  end
end

ruby_block 'make sure we really reproed' do
  block do
    if arch_mode_32_bit_failed == true && arch_mode_64_bit_failed == false
      puts "\n\n****Repro worked -- file is only detected by 64-bit redirection workaround and is invisible in default 32 bit mode"
    else
      raise 'The repro failed and we may still not understand this issue, or you have a 64-bit Ruby + Chef :)'
    end
  end
end
```

**Comment by adamedx** [comment]
@btm, @jdmundrawala, @ksubrama, here's a repro for Sean's issue yesterday. Kartik's hypothesis is correct, as proven by the sample recipe above.

tl;dr -- we need to go 64-bit.

**Comment by smurawski** [comment]
The other workaround is to use the built in workarounds for the redirection - 

```
file 'c:/hello..txt' do
  content 'world'
  only_if { ::File.exist? 'c:/windows/sysnative/tasks/somefile' }
end
```

For program files, there are the environmental variables ProgramFiles(x86) and ProgramW6432.

**Comment by adamedx** [comment]
Good point @smurawski -- does that also work on 32-bit? That's my only caveat with good old sysnative, I can't remember if 32-bit windows is ok with that or not.

**Comment by smurawski** [comment]
It is not.  However, 2008 R2 is the last version of server to offer a 32 bit install.

We could just offer a helper method that checks the system bitness and fixes file paths based on that.

**Comment by adamedx** [comment]
The current solution to address this is to provide a Chef package for Windows that uses a 64-bit Ruby (i.e. 64-bit procesess) and will not have this issue.

**Comment by btm** [comment]
As of Chef 12.9, released on 2016-04-13, we have 64-bit Windows builds. I don't see a strong case for any workarounds or accommodations for 32-bit builds related to file redirection, as I don't see a reason for users to not move toward 64-bit builds.

https://discourse.chef.io/t/chef-client-12-9-released/8232
https://blog.chef.io/2015/11/19/pre-release-announcement-native-64-bit-windows-chef-client/
https://blog.chef.io/2016/02/08/update-on-native-64-bit-windows-chef-client/
